bpf: Remove CT_REOPENED, return CT_NEW instead

jrajahalme opened the pull request:
CT_REOPENED was originally added in https://github.com/cilium/cilium/pull/13340 to emit policy verdicts for apparently re-opened TCP connections, which are in fact more likely to be newly opened TCP connections rather than re-opened ones, as the CT entries may live minutes after the TCP state from the endpoints has already timed out.

This added complexity to call sites, forcing differentiation between CT_NEW and CT_REOPENED. In all cases some CT entry field values were left stale, e.g., 'proxy_redirect' after a policy change.

Instead of adjusting each call site to behave properly for CT_REOPENED, return CT_NEW instead, and make the observable CT lookup behavior the same as for CT_NEW in that case, most notably by not updating the passed in `*ct_state'.

This change fixes proxy redirection bug where return packets are not redirected to an L7 proxy when (a stale) CT entry is missing the 'proxy_redirect' flag.

For this bug to trigger a pod needs to open a new TCP connection using the same (ephemeral) source port to the same destination before and after a change in policy adding an L7 policy applicable to that connection. While this is a bug that has surfaced in CI, this bug is less likely to be hit when policies are more stable. Given this we need further discussion about backporting this bug fix to the stable releases.

Fixes: #27762
Fixes: #13340

Note: This PR is an alternative to https://github.com/cilium/cilium/pull/32614, only one of these should be merged.

```release-note
Datapath conntrack entries for reopened connections are fully reinitialized to fix rare L7 proxy redirect failures.
```

jrajahalme committed:
datapath: Deprecate trace reason REOPENED

Datapath no longer returns a trace reason for REOPENED. Keep the Go
symbol for compatibility with older datapaths, but rename it to mark it
as deperecated.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

julianwiedmann commented on bpf/lib/conntrack.h:
```diff
@@ -320,7 +321,11 @@ __ct_lookup(const void *map, struct __ctx_buff *ctx, const void *tuple,
 				entry->seen_non_syn = false;
 
 				*monitor = ct_update_timeout(entry, is_tcp, dir, seen_flags);
-				return CT_REOPENED;
+
+				/* Return CT_NEW so that the caller creates a new entry instead of
+				 * updating the old one. (For policy drops the old entry remains.)
+				 */
+				return CT_NEW;
```
Do you think we should remove the `ct_reset_closing()` and `ct_reset_seen_flags()` here as well? I'm undecided on the `ct_update_timeout()` call ... it almost feels like we should just have a `goto ct_new`.

jrajahalme replied:
Not sure. It is still possible that a new entry will not be created due to a policy DROP, so I left these be.

julianwiedmann replied:
Hmm yes makes sense.

julianwiedmann commented on bpf/bpf_lxc.c:
```diff
@@ -1979,28 +1961,11 @@ ipv4_policy(struct __ctx_buff *ctx, struct iphdr *ip4, int ifindex, __u32 src_la
-				ct_update_nodeport(get_ct_map4(tuple), tuple,
-						   node_port);
-		}
-	}
```
https://github.com/cilium/cilium/pull/32642 already landed, so this will need a trivial rebase.

jrajahalme replied:
Done!

julianwiedmann commented on bpf/lib/nodeport.h:
```diff
@@ -2106,10 +2105,8 @@ nodeport_extract_dsr_v4(struct __ctx_buff *ctx,
 	 */
-	if (tuple->nexthdr == IPPROTO_TCP)
-		ct_update_dsr(get_ct_map4(&tmp), &tmp, false);
 
```
I'm not 100% on this. Right now it covers not just `CT_REOPENED`, but also the case where we receive multiple SYNs (by a oddly-behaving client :)). The 2nd/3rd/... SYN would get matched as `CT_ESTABLISHED`, no? So we still want to update the DSR-state accordingly.

jrajahalme replied:
This was not done for `CT_ESTABLISHED`, so this change is correct IMO. Also, I don't know if it is possible for dsr to be `true` in the case you refer to (re-transmitted `SYN`s).

julianwiedmann replied:
>This was not done for CT_ESTABLISHED, so this change is correct IMO.

I don't follow, could you elaborate?

>Also, I don't know if it is possible for dsr to be true in the case you refer to (re-transmitted SYNs).

How do you mean? Yes, I'm pretty sure the described scenario (first SYN with DSR-info, second SYN without DSR-info) is possible.

jrajahalme replied:
You are right, I had this mixed up. Reverted this change (also for v6).

kaworu commented on api/v1/flow/flow.proto:
```diff
@@ -196,7 +196,7 @@ enum TraceReason {
     REPLY = 3;
     RELATED = 4;
-    REOPENED = 5;
+    DEPRECATED_REOPENED = 5;
```
I'm fine with internal renamings for clarity, but this one would be a breaking change because enums are represented with their text value (i.e. not their protobuf field number) when marshaled into JSON according to our configuration.
```suggestion
    REOPENED = 5 [deprecated=true];
```

jrajahalme replied:
Did not realize this, thanks! Reverted this change, except for `[deprecated=true]` you suggested and the Go const name.

kaworu reviewed (approved):
Thanks for the update, Hubble changes LGTM now.

jrajahalme committed:
bpf: Remove CT_REOPENED, return CT_NEW instead

CT_REOPENED was originally added in
https://github.com/cilium/cilium/pull/13340 to emit policy verdicts for
apparently re-opened TCP connections, which are in fact more likely to be
newly opened TCP connections rather than re-opened ones, as the CT
entries may live minutes after the TCP state from the endpoints has
already timed out.

This added complexity to call sites, forcing differentiation between
CT_NEW and CT_REOPENED. In all cases some CT entry field values were left
stale, e.g., 'proxy_redirect' after a policy change.

Instead of adjusting each call site to behave properly for CT_REOPENED,
return CT_NEW instead, and make the observable CT lookup behavior the
same as for CT_NEW in that case, most notably by not updating the passed
in `*ct_state'.

This change fixes proxy redirection bug where return packets are not
redirected to an L7 proxy when (a stale) CT entry is missing the
'proxy_redirect' flag.

Fixes: #27762
Fixes: #13340
Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

julianwiedmann reviewed (approved):
Looks good, thank you Jarno!

:heavy_check_mark:, assuming that `ct_update_dsr()` is re-instated to fix the build.
